### DISCUSSION TITLE
FastDDS and OpenDDS interoperability issue

### DISCUSSION THREAD

**Discussion by husnaini798** [eProsima/Fast-DDS#3609] [Unofficial support]
### Is there an already existing issue for this?

- [X] I have searched the existing issues

### Expected behavior

Note: Security feature disabled

| FastDDS    | OpenDDS | Results |
| ----------- | ----------- |----------- |
| Publisher     | Subscriber|  works fine|
| Subscriber |  Publisher|  should work fine but doesn't|

### Current behavior

Note: Security feature disabled

| FastDDS    | OpenDDS | Results |
| ----------- | ----------- |----------- |
| Publisher     | Subscriber|  works fine|
| Subscriber |  Publisher| Subscriber does get the call to 'on_subscription_matched' callback, when Publisher  <br>starts publishing. But subscriber never gets the call to 'on_data_available' callback |

### Steps to reproduce

FastDDS's SecureHelloWorldExample extended and with security code commented out. Also set the default QOS for participant, topic, reader and writer

For OpenDDS, used the following code for testing the OpenDDS code:
https://github.com/omg-dds/dds-security
Also commented out the security code in here. And set the default QOS for participant, topic, reader and writer

The two zipped helloWorld code folders (for OpenDDS and FastDDS) are only 80Kb, Please let me know how to share those.

Thanks

### Fast DDS version/commit

version 2.8.1

### Platform/Architecture

Ubuntu Focal 20.04 amd64

### Transport layer

Default configuration, UDPv4 & SHM

### Additional context

IDL used on both OpenDDS and FastDDS
`module Messenger {
struct HelloWorld
{
	unsigned long index;
	string message;
};
};
`

**Comment by husnaini798** [comment]
[github URL for the two zip files containing the OpenDDS and FastDDS code](https://github.com/husnaini798/DDSInterop)

**Comment by JLBuenoLopez** [comment]
Hi @husnaini798,

Have you tested against other DDS implementations? Fast DDS release testing checks interoperability with several vendors. We are aware about the issue you are reporting but we have also reproduced the same issue with at least one other vendor so we assume that the interoperability issue might be at OpenDDS side.

**Comment by husnaini798** [comment]
Hi Jose,
Thanks for confirming our findings.

 Not yet. I was thinking of trying Cyclone DDS next, with both FastDDS and
OpenDDS.

Would you be able to inform me which other vendor had the same issue?

Thanks

**Comment by Mario-DL** [comment]
According to the [CONTRIBUTING.md](https://github.com/eProsima/Fast-DDS/blob/master/CONTRIBUTING.md) I am moving this to the proper discussion forum.

**Comment by andreclerigo** [comment]
Hello @husnaini798, could you share more information on your findings?

**Comment by EduPonz** [comment]
Hi everyone,

Since [Fast DDS v2.13.0](https://github.com/eProsima/Fast-DDS/releases/tag/v2.13.0) it is possible to configure the [DataRepresentationQosPolicy](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/core/policy/xtypesExtensions.html#datarepresentationqospolicy) for both DataReaders and DataWriters. Setting the ID to [XCDR2_DATA_REPRESENTATION](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/core/policy/datarepresentationid.html#_CPPv4N8eprosima7fastdds3dds20DataRepresentationId25XCDR2_DATA_REPRESENTATIONE) in your DataWriters would make Fast DDS and OpenDDS to interoperate. You can verify this by using [Fast DDS Shapes Demo](https://eprosima-shapes-demo.readthedocs.io/en/latest/first_steps/first_steps.html#publishing-a-topic-shape) and Open DDS's iShape.

**Comment by kshang** [comment]
@husnaini798 你用的FastDDS和Cyclonedds分别是什么版本啊？我使用了security功能后，FastDDS和Cyclonedds无法通信了。

**Comment by vidalperezbohoyo** [comment]
Same issue in 2024. Only working in one direction. I think that data representation is not the problem. Using FastDDS 2.14 and OpenDDS 3.30. Same Qos and XCDR2. In FastDDS data reader in this version, qos didn't have representation member so i put XCDR2 in topic. The FastDDS 3.1.0 version has representation in DataReaderQos. But i forced to use 2.14. I don't known the difference between DataReaderQos and ReaderQos. I give DataReaderQos (without representation field in 2.14) to data reader constructor.
The only callback is called is on subscription matched. OpenDDS with OpenDDS working fine with this idl and data. And FastDDS with FastDDS same.

**Reply by JonathanDev22** [reply]
I have the same problem, I only have communication in one direction and I need to process this data, have you found the solution?

**Reply by cferreiragonz** [reply]
Hello @vidalperezbohoyo and @JonathanDev22. 

If you take a look at the test report of the OMG-DDS (it can be found [here](https://github.com/omg-dds/dds-rtps/actions/runs/11555040003)) there should be no problem between Fast DDS and Open DDS interoperability. Also, we internally tested that Fast DDS 2.14 is able to communicate with Open DDS 3.30 using ShapesDemo and the equivalent version of Open DDS. Everything worked fine.

Could you test if the same issue occurs with basic types such as the ones used by ShapesDemo?
